flink-libraries/flink-table/src/test/scala/org/apache/flink/table/formats/TableFormatFactoryServiceTest.scala

```diff
@@ -0,0 +1,122 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.formats
+
+import java.util.{HashMap => JHashMap, Map => JMap}
+
+import org.apache.flink.table.api.{AmbiguousTableFormatException, NoMatchingTableFormatException}
+import org.apache.flink.table.formats.utils.TestAmbiguousTableFormatFactory
+import org.junit.Assert.{assertNotNull, assertTrue}
+import org.junit.Test
+
+/**
+  * Tests for [[TableFormatFactoryService]].
+  */
+class TableFormatFactoryServiceTest {
+
+  @Test
+  def testValidProperties(): Unit = {
+    val props = properties()
+    assertNotNull(
+      TableFormatFactoryService.find(classOf[TableFormatFactory[_]], props))
+  }
+
+  @Test
+  def testDifferentContextVersion(): Unit = {
+    val props = properties()
+    props.put("format.property-version", "2")
+    // the format should still be found
+    assertNotNull(TableFormatFactoryService.find(classOf[TableFormatFactory[_]], props))
+  }
+
+  @Test
+  def testAmbiguousMoreSupportSelection(): Unit = {
+    val props = properties()
+    props.remove("format.important")
+    props.put("format.special_path", "/what/ever")
+    assertTrue(
+      TableFormatFactoryService
+        .find(classOf[TableFormatFactory[_]], props)
+        .isInstanceOf[TestAmbiguousTableFormatFactory])
+  }
+
+  @Test
+  def testAmbiguousClassBasedSelection(): Unit = {
+    val props = properties()
+    props.remove("format.important")
+    assertTrue(
+      TableFormatFactoryService
+        .find(classOf[TestAmbiguousTableFormatFactory], props)
+        .isInstanceOf[TestAmbiguousTableFormatFactory])
+  }
+
+  @Test
+  def testAmbiguousSchemaBasedSelection(): Unit = {
+    val props = properties()
+    props.remove("format.important")
+    props.put("schema.weird_field", "unknown") // this is unknown to the schema derivation factory
+    assertTrue(
+      TableFormatFactoryService
+        .find(classOf[TableFormatFactory[_]], props)
+        .isInstanceOf[TestAmbiguousTableFormatFactory])
+  }
+
+  @Test(expected = classOf[NoMatchingTableFormatException])
+  def testMissingClass(): Unit = {
+    val props = properties()
+    // this class is not a valid factory
+    TableFormatFactoryService.find(classOf[TableFormatFactoryServiceTest], props)
+  }
+
+  @Test(expected = classOf[NoMatchingTableFormatException])
+  def testInvalidContext(): Unit = {
+    val props = properties()
+    props.put("format.type", "FAIL") // no context specifies this
+    TableFormatFactoryService.find(classOf[TableFormatFactory[_]], props)
+  }
+
+  @Test(expected = classOf[NoMatchingTableFormatException])
+  def testUnsupportedProperty(): Unit = {
+    val props = properties()
+    props.put("format.path_new", "/new/path") // no factory has this
+    TableFormatFactoryService.find(classOf[TableFormatFactory[_]], props)
+  }
+
+  @Test(expected = classOf[AmbiguousTableFormatException])
+  def testAmbiguousFactory(): Unit = {
```

Comment:
how does this differ from `testAmbiguousSchemaBasedSelection`?

Reply:
I added a comment to the other test.